magma/compile_guard.py

```diff
@@ -1,40 +1,79 @@
 import contextlib
-import dataclasses
 from typing import Optional, Tuple
 
 from magma.bits import BitsMeta
-from magma.clock import Clock
+from magma.clock import ClockTypes
 from magma.circuit import Circuit, CircuitBuilder
 from magma.conversions import as_bits
 from magma.definition_context import definition_context_manager
 from magma.digital import DigitalMeta
 from magma.generator import Generator2
 from magma.inline_verilog import inline_verilog
 from magma.interface import IO
+from magma.logging import root_logger
+from magma.passes.group import GrouperBase, InstanceCollection
 from magma.primitives.mux import infer_mux_type
-from magma.ref import AnonRef, ArrayRef, DefnRef, InstRef, TupleRef
-from magma.t import Kind, In, Out
+from magma.ref import InstRef, get_ref_defn, get_ref_inst
+from magma.t import Kind, Type, In, Out
 from magma.type_utils import type_to_sanitized_string
-from magma.value_utils import make_selector
 
 
-def _get_top_ref(ref):
-    if isinstance(ref, TupleRef):
-        # TODO(rsetaluri): Fix port name collision.
-        return _get_top_ref(ref.tuple.name)
-    if isinstance(ref, ArrayRef):
-        # TODO(rsetaluri): Fix port name collision.
-        return _get_top_ref(ref.array.name)
-    return ref
+_logger = root_logger().getChild("compile_guard")
 
 
-@dataclasses.dataclass(frozen=True)
-class _CompileGuardState:
-    ckt: CircuitBuilder
-    ctx_mgr: contextlib.AbstractContextManager
+class _Grouper(GrouperBase):
+    def __init__(self, instances: InstanceCollection, builder: CircuitBuilder):
+        super().__init__(instances)
+        self._builder = builder
+        self._port_index = 0
+        self._clock_types = set()
+
+    def _visit_input_connection(self, _: Type, drivee: Type):
+        # NOTE(rsetaluri): Since the driver might be traced (i.e. not an
+        # immediate driver), we disregard it and grab the immediate driver
+        # (value() vs. trace()).
+        driver = drivee.value()
+        new_port = self._builder.add_port(In(type(driver)))
+        drivee.unwire(driver)
+        drivee @= new_port
+        external = getattr(self._builder, new_port.name.name)
+        external @= driver
+
+    def _visit_output_connection(self, driver: Type, drivee: Type):
+        new_port = self._builder.add_port(Out(type(driver)))
+        new_port @= driver
+        external = getattr(self._builder, new_port.name.name)
+        old_driver = drivee.value()
+        drivee.unwire(old_driver)
+        new_driver = external
+        drivee @= new_driver
```

Comment:
We could consider adding an a "rewire" API, something like:

```python
drivee.rewire(new_driver)
```

which would suppress the rewiring warning by internal doing the unwire call.